lib/interactive_key.rb

```diff
@@ -0,0 +1,553 @@
+  attr_accessor :language_to_use
+
+  #list of descriptors reduced by keyword_ids
+  attr_accessor :descriptors_with_filter
+
+  #list of rows to be included into the matrix
+  attr_accessor :rows_with_filter
+
+  #return the list of descriptors with selections
+  #  attr_accessor :used_descriptors
+
+  #return the list of useful descriptors
+  #  attr_accessor :useful_descriptors
+
+  #return the list of descriptors not useful for identification
+  #  attr_accessor :not_useful_descriptors
+
+  #return the list of descriptors and thair states
+  attr_accessor :list_of_descriptors
+
+  #list of remaining rows
+  attr_accessor :remaining
+
+  #list of eliminated rows
+  attr_accessor :eliminated
+
+  #hash of used descriptors and their states
+  attr_accessor :selected_descriptors_hash
+
+  #temporary hash of rows; used for calculation of remaining and eliminated rows
+  attr_accessor :row_hash
+
+  #temporary hash of descriptors; used for calculation of useful and not useful characters and their states
+  attr_accessor :descriptors_hash
+
+  def initialize(
+    observation_matrix_id: nil,
+    project_id: nil,
+    language_id: nil,
+    keyword_ids: nil,
+    row_filter: nil,
+    sorting: 'optimized',
+    error_tolerance: 0,
+    identified_to_rank: nil,
+    eliminate_unknown: nil,
+    selected_descriptors: nil)
+
+    # raise if observation_matrix_id.blank? || project_id.blank?
+    @observation_matrix_id = observation_matrix_id
+    @project_id = project_id
+
+    # This is nothing?   
+    # @observation_matrix = observation_matrix
+
+    # Do you mean
+    @observation_matrix = ObservationMatrix.where(project_id: project_id).find(observation_matrix_id)
+    @observation_matrix_citation = @observation_matrix.source
+    @descriptor_available_languages = descriptor_available_languages
+    @language_id = language_id
+    @language_to_use = language_to_use
+    @keyword_ids = keyword_ids
+    @descriptor_available_keywords = descriptor_available_keywords
+    @descriptors_with_filter = descriptors_with_keywords
+    @row_filter = row_filter
+    @rows_with_filter = get_rows_with_filter
+    @sorting = sorting
+    @error_tolerance = error_tolerance.to_i
+    @eliminate_unknown = eliminate_unknown == 'true' ? true : false
+    @identified_to_rank = identified_to_rank
+    @selected_descriptors = selected_descriptors
+    @selected_descriptors_hash = selected_descriptors_hash_initiate
+    @row_hash = row_hash_initiate
+    @descriptors_hash = descriptors_hash_initiate
+    #main_logic
+    @remaining = remaining_taxa
+    @eliminated = eliminated_taxa
+
+    @list_of_descriptors = useful_descriptors
+
+    @row_hash = nil
+    @descriptors_hash = nil
+    @rows_with_filter = nil
+    @descriptors_with_filter = nil
+  end
+
+  def observation_matrix
+    ObservationMatrix.where(id: @observation_matrix_id, project_id: @project_id).first
+  end
+
+  def descriptors
+    if @sorting == 'weighted'
```

Comment:
Don't use the variable version (ampersand) of an attribute use its getter, e.g. `@sorting` -> `sorting`.   Only use the `@` version when setting/changing it.

 Change throughout.